contrib/zookeeper/piped/contrib/zookeeper/providers.py

```
@@ -0,0 +1,201 @@
+            if key not in self.possible_events:
+                e_msg = 'Invalid event: {0}.'.format(key)
+                detail = 'Use one of the possible events: {0}'.format(self.possible_events)
+                raise exceptions.ConfigurationError(e_msg, detail)
+
+            self.events[key] = dict(provider='pipeline.{0}'.format(value))
+        
+        self.dependencies = runtime_environment.create_dependency_map(self, **self.events)
+
+    def _started(self):
+        self.connecting = None
+        self.on_connected(self)
+        self._on_event('connected')
+
+    @defer.inlineCallbacks
+    def _on_event(self, event_name):
+        baton = dict(event=event_name, client=self)
+
+        try:
+            pipeline = yield self.dependencies.wait_for_resource(event_name)
+            yield pipeline.process(baton)
+        except KeyError as ae:
+            # we have no pipeline for this event
+            pass
+
+    @defer.inlineCallbacks
+    def _watch_connection(self, client, event):
+        # see client.STATE_NAME_MAPPING for possible values for event.state_name
+        if event.state_name == 'connected':
+            self.on_connected(self)
+            self._on_event('reconnected')
+        elif event.state_name == 'connecting':
+            # TODO: if we're in "connecting" for too long, give up and give us a new connection (old session might be bad
+            #       because the server rollbacked -- see https://issues.apache.org/jira/browse/ZOOKEEPER-832)
+            self.on_disconnected(event.state_name)
+            self._on_event('reconnecting')
+        elif event.state_name == 'expired':
+            self.on_disconnected(event.state_name)
+            self._on_event('expired')
+            # force a full reconnect in order to ensure we get a new session
+            yield self.stopService()
+            yield self.startService()
+        else:
+            # do nothing for auth failed or associating
```

Silently ignoring authentication failures smells like a frustration source.